File: JavaScript/739.每日温度.js
/*
 * @lc app=leetcode.cn id=739 lang=javascript
 *
 * [739] 每日温度
 */

// @lc code=start
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
/* 
Accepted
47/47 cases passed (256 ms)
Your runtime beats 58.15 % of javascript submissions
Your memory usage beats 6.38 % of javascript submissions (59.8 MB)
*/
var dailyTemperatures = function (temperatures) {
  // 存放的是当前温度所在temperatures中的序号
  const stack = []
  const len = temperatures.length
  const arr = new Array(len).fill(0)

  for (let i = 0; i < temperatures.length; i++) {
    const t = temperatures[i]
    // 当栈中有数，并且当前温度大于栈中最后一位温度的时候，将栈一次出栈，并记录
    while (stack.length && temperatures[i] > temperatures[stack[stack.length - 1]]) {
      const index = stack.pop()
      arr[index] = i - index
    }
    stack.push(i)
  }

  return arr
};
// @lc code=end

File: Python3/739.每日温度.py
#
# @lc app=leetcode.cn id=739 lang=python3
#
# [739] 每日温度
#

# @lc code=start
"""
暴力枚举
没有问题但会超时，以前好像不会，leetcode垃圾
"""

"""
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        for idx, t in enumerate(temperatures):
            for i, te in enumerate(temperatures[idx + 1:]):
                if te > t:
                    res[idx] = i + 1
                    break

        return res
"""

"""
单调栈
Accepted
47/47 cases passed (192 ms)
Your runtime beats 62.19 % of python3 submissions
Your memory usage beats 96.43 % of python3 submissions (19.7 MB)
"""

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        stack = []

        for idx in range(len(temperatures) - 1, -1, -1):
            while stack and temperatures[idx] >= temperatures[stack[-1]]:
                stack.pop()
            if stack:
                res[idx] = stack[-1] - idx
            stack.append(idx)

        return res

# @lc code=end
